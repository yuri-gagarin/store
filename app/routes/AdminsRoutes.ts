import passport from "passport";
import AdminsController from "../controllers/admins_controller/AdminsController";
import { RouteConstructor } from "./helpers/routeInterfaces";

import  { Router } from "express";

class AdminRoutes extends RouteConstructor<AdminsController> {
  private registerAdminRoute = "/api/admins/register";
  private updateAdminRoute = "/api/admins/update/:adminId";
  private deleteAdminRoute = "/api/admins/delete/:adminId";
  private loginAdminRoute = "/api/admins/login";
  private logoutAdminRoute = "/api/admins/logout"

  constructor(router: Router, controller: AdminsController) {
    super(router, controller)
    this.initializeRoutes();
  }
  protected initializeRoutes(): void {
    this.registerAdmin();
    this.updateAdmin();
    this.deleteAdmin();
    this.deleteAdmin();
    this.loginAdmin();
    this.logoutAdmin();
  }
  private registerAdmin (): void {
    this.Router.route(this.registerAdminRoute).post(this.controller.register);
  }
  private updateAdmin (): void {
    this.Router.route(this.updateAdminRoute).patch(this.controller.editRegistration);
  }
  private deleteAdmin (): void {
    this.Router.route(this.deleteAdminRoute).delete(passport.authenticate("adminJWT", { session: false }), this.controller.deleteRegistration);
  }
  private loginAdmin (): void {
    this.Router.route(this.loginAdminRoute).post(this.controller.login);
  }
  private logoutAdmin (): void {
    this.Router.route(this.logoutAdminRoute).delete(this.controller.logout);
  }
};

export default AdminRoutes;