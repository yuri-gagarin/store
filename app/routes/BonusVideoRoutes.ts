import { Router} from "express";
import { RouteConstructor } from "./helpers/routeInterfaces";
import { IGenericController } from "../controllers/helpers/controllerInterfaces";

class BonusVideoRoutes extends RouteConstructor<IGenericController> {
  private viewAllBonusVideosRoute = "/api/bonus_videos";
  private viewBonusVideoRoute = "/api/bonus_videos/:_id";
  private createBonusVideoRoute = "/api/bonus_videos/create";
  private editBonusVideoRoute = "/api/bonus_videos/update/:_id";
  private deleteBonusVideoRoute = "/api/bonus_videos/delete/:_id";
  
  constructor (router: Router, controller: IGenericController) {
    super(router, controller);
    this.initializeRoutes();
  }
  protected initializeRoutes (): void {
    this.getAllBonusVideos();
    this.getBonusVideo();
    this.createBonusVideo();
    this.editBonusVideo();
    this.deleteBonusVideo();
  }
  private getAllBonusVideos (): void {
    this.Router.route(this.viewAllBonusVideosRoute).get(this.controller.getMany);
  }
  private getBonusVideo (): void {
    this.Router.route(this.viewBonusVideoRoute).get(this.controller.getOne);
  }
  private createBonusVideo (): void {
    this.Router.route(this.createBonusVideoRoute).post(this.controller.create);
  }
  private editBonusVideo (): void {
    this.Router.route(this.editBonusVideoRoute).patch(this.controller.edit);
  }
  private deleteBonusVideo (): void {
    this.Router.route(this.deleteBonusVideoRoute).delete(this.controller.delete);
  }
}

export default BonusVideoRoutes;