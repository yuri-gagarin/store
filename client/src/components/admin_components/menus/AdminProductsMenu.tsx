import React, { useState, useEffect, useRef } from "react";
import { Menu, MenuItemProps } from "semantic-ui-react";
// routing //
import { withRouter, RouteComponentProps, useRouteMatch } from "react-router-dom"
// css imports //
import "./css/adminProductsMenu.css";
// actions and state //
import { AppAction } from "../../../state/Store";

interface Props extends RouteComponentProps {
  dispatch: React.Dispatch<AppAction>
}
const AdminProductsMenu: React.FC<Props> = ({ history, location, dispatch }): JSX.Element => {
  const [ scrolled, setScrolled ] = useState<boolean>(false);
  const [ activeItem, setActiveItem ] = useState<string>("view_all");
  const [ menuOpen, setMenuOpen ] = useState<boolean>(false);

  const match = useRouteMatch("/admin/home/my_products");
  const adminProductsMenuRef = useRef<HTMLDivElement>(document.createElement("div"));

  const handleItemClick = (e: React.MouseEvent, { name }: MenuItemProps): void => {
    setActiveItem(String(name));


    switch (name) {
      case "view_all": {
        history.push(match?.path + "/all");
        break;
      }
      case "create": {
        history.push(match?.path + "/create");
        dispatch({ type: "CLEAR_CURRENT_PRODUCT", payload: null });
        break;
      }
      case "manage": {
        history.push(match?.path + "/manage");
        break;
      }
      case 'sorted': {
        history.push(match?.path + "/view_sorted");
        break;
      }
      default: history.push("/admin/home");
    }
  }
 
  const scrollListener = () => {
    if (window.scrollY > 1) {
      setScrolled(true);
    } else if (window.scrollY === 0) {
      setScrolled(false)
    } 
  };
  // lifecycle hooks //
  useEffect(() => {
    const currentURL = location.pathname;
    setTimeout(() => {
      setMenuOpen(true);
    }, 200);
    if (currentURL.match(/all/)) {
      setActiveItem("view_all");
    } else if (currentURL.match(/create/)) {
      setActiveItem("create");
    } else if (currentURL.match(/manage/)) {
      setActiveItem("manage");
    } else if (currentURL.match(/sorted/)) {
      setActiveItem("sorted");
    }
    // menu animation timeout //
    setTimeout(() => {
      setMenuOpen(true);
    }, 200);
  }, []);

  useEffect(() => {

    window.addEventListener("scroll", scrollListener, true);
    return () => window.removeEventListener("scroll", scrollListener);

  }, [adminProductsMenuRef]);

  return (
    <div className={ scrolled ? "adminProductsMenuFixed menuScrolled" : "adminProductsMenuFixed"} ref={adminProductsMenuRef}>
      <Menu tabular className={ menuOpen ? "adminProductsMenu productsMenuOpen" : "adminProductsMenu" }>
      <Menu.Item
          name='view_all'
          content="View All Products"
          active={activeItem === 'view_all'}
          onClick={handleItemClick}
        />
        <Menu.Item
          name='create'
          content="Create New Product"
          active={activeItem === 'create'}
          onClick={handleItemClick}
        />
        <Menu.Item
          name='manage'
          content="Manage Products"
          active={activeItem === 'manage'}
          onClick={handleItemClick}
        />
        <Menu.Item
          name="sorted"
          content="View Sorted Products"
          active={activeItem === "sorted"}
          onClick={handleItemClick}
        >
        </Menu.Item>
      </Menu>
    </div>
  )
};

export default withRouter(AdminProductsMenu);