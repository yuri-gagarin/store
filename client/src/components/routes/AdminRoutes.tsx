import React from "react";
import { Route } from "react-router-dom";
// component imports //
import LoginComponent from "../admin_components/login/LoginComponent";

const AdminRoutes: React.FC<{}> = (props): JSX.Element =>  {
  return (
    <React.Fragment>
      <Route exact path="/admin/login">
        <LoginComponent />
      </Route>
      
    </React.Fragment>
    
  );
};

export default AdminRoutes;