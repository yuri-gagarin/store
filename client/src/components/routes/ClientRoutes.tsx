import React from "react";
import { Switch, Route } from "react-router-dom";
// component imports //
import HomeScreenComponent from "../client_components/home_screen/HomeScreenComponent";
import StoreHolder from "../client_components/store_screen/StoreHolder";
import MainNavSideBar from "../client_components/navbars/MainNavSidebar";

const ClientRoutes: React.FC<{}> = (props): JSX.Element =>  {
  return (
    <MainNavSideBar>
      <Switch>
        <Route exact path="/">
          <HomeScreenComponent title=""/>
        </Route>
        <Route exact path="/store">
          <StoreHolder />
        </Route>
      </Switch>
    </MainNavSideBar>
    
  );
};

export default ClientRoutes;