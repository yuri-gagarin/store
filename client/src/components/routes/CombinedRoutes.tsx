import React from "react";
import { BrowserRouter, Route, Switch } from "react-router-dom";
import MainNavSideBar from "../client_components/navbars/MainNavSidebar";
import LoginComponent from "../admin_components/login/LoginComponent";
import AdminHomeScreen from "../admin_components/home/AdminHomeScreen";
// import AdminRoutes from "./AdminRoutes";
// component imports //

const CombinedRoutes: React.FC<{}> = (props): JSX.Element =>  {
  return (
    <BrowserRouter>
      <Switch>
        <Route path="/admin/login">
          <LoginComponent />
        </Route>
        <Route path="/admin/home">
          <AdminHomeScreen />
        </Route>
        <Route path="/">
          <MainNavSideBar />
        </Route>
      </Switch>
      
      
    </BrowserRouter>
  );
};

export default CombinedRoutes;