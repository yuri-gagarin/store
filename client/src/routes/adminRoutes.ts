export const AdminProductRoutes = {
  HOME_ROUTE: "/admin/home/my_products",
  VIEW_ALL_ROUTE: "/admin/home/my_products/view_all",
  CREATE_ROUTE: "/admin/home/my_products/create",
  MANAGE_ROUTE: "/admin/home/my_products/manage",
  SHOW_ROUTE: "/admin/home/my_products/show",
  VIEW_SORTED_ROUTE: "/admin/home/my_products/view_sorted"
};

export const AdminServiceRoutes = {
  ADMIN_SERVICES_HOME_ROUTE: "/admin/home/my_services",
  ADMIN_SERVICES_VIEW_ALL_ROUTE: "/admin/home/my_services/view_all",
  ADMIN_SERVICES_CREATE_ROUTE: "/admin/home/my_services/create",
  ADMIN_SERVICES_MANAGE_ROUTE: "/admin/home/my_services/manage",
  ADMIN_SERVICES_SHOW_ROUTE: "/admin/home/my_services/show",
  ADMIN_SERVICES_VIEW_SORTED_ROUTE: "/admin/home/my_services/view_sorted"
};

export const AdminStoreRoutes = {
  ADMIN_STORES_HOME_ROUTE: "/admin/home/my_stores",
  ADMIN_STORES_CREATE_ROUTE: "/admin/home/my_stores/create",
  ADMIN_STORES_MANAGE_ROUTE: "/admin/home/my_stores/manage",
  ADMIN_STORES_SHOW_ROUTE: "/admin/home/my_stores/show",
  ADMIN_STORES_VIEW_ALL_ROUTE: "/admin/home/my_stores/view_all"
};

export const AdminStoreItemRoutes = {
  HOME_ROUTE: "/admin/home/my_store_items",
  CREATE_ROUTE: "/admin/home/my_store_items/create",
  MANAGE_ROUTE: "/admin/home/my_store_items/manage",
  SHOW_ROUTE: "/admin/home/my_store_items/show",
  VIEW_ALL_ROUTE: "/admin/home/my_store_items/view_all",
  VIEW_SORTED_ROUTE: "/admin/home/my_store_items/view_sorted"
};

