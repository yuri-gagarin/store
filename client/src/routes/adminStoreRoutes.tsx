export default {
  ADMIN_STORES_HOME_ROUTE: "/admin/home/my_stores",
  ADMIN_STORES_CREATE_ROUTE: "/admin/home/my_stores/create",
  ADMIN_STORES_MANAGE_ROUTE: "/admin/home/my_stores/manage",
  ADMIN_STORES_SHOW_ROUTE: "/admin/home/my_stores/show",
  ADMIN_STORES_VIEW_ALL_ROUTE: "/admin/home/my_stores/view_all"
}