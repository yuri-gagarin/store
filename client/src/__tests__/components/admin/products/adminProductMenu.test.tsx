import React from "react";
import { Menu } from "semantic-ui-react";
// test dependencies
import { mount, ReactWrapper } from "enzyme";
import { MemoryRouter, Router } from "react-router-dom";
// component imports //
import AdminProductsMenu from "../../../../components/admin_components/menus/AdminProductsMenu";
// component dependencies //
import { AdminProductRoutes } from "../../../../routes/adminRoutes";

describe("AdminProductMenu component render tests", () => {
  let wrapper: ReactWrapper;
  beforeAll(() => {
    wrapper = mount(
      <MemoryRouter initialEntries={[AdminProductRoutes.HOME_ROUTE]} keyLength={0}>
        <AdminProductsMenu dispatch={jest.fn}/>
      </MemoryRouter>
    )
  });
  describe("Default component render", () => {
    it("Should properly render and match snapshot", () => {
      expect(wrapper.find(AdminProductsMenu)).toMatchSnapshot();
    });
    it("Should have 4 'Menu.Item' components", () => {
      const menuItems = wrapper.find(Menu.Item);
      expect(menuItems.length).toEqual(4);
    });
    it("Should not have any of the links in 'active' state", () => {
      const menuItems = wrapper.find(Menu.Item);
      menuItems.forEach((item) => {
        expect(item.props().active).toEqual(false);
      })
    })
  });

  describe("AdminProductMenu component nav button tests", () => {
    
    it("Should correctly respond to 'View All' 'Menu.Item' click", () => {
      window.scrollTo = jest.fn;
      const menuItems = wrapper.find(Menu.Item)
      menuItems.at(0).simulate("click");
      wrapper.update();
      // assert correct rendering //
      expect(wrapper.find(Menu.Item).at(0).props().active).toEqual(true);
      expect(wrapper.find(Menu.Item).at(1).props().active).toEqual(false);
      expect(wrapper.find(Menu.Item).at(2).props().active).toEqual(false);
      expect(wrapper.find(Menu.Item).at(3).props().active).toEqual(false);
      // assert correct client route //
      expect(wrapper.find(Router).props().history.location.pathname).toEqual(AdminProductRoutes.VIEW_ALL_ROUTE);

    });
    it("Should correctly respond to 'Create' 'Menu.Item' click", () => {
      window.scrollTo = jest.fn;
      const menuItems = wrapper.find(Menu.Item)
      menuItems.at(1).simulate("click");
      wrapper.update();
      // assert correct rendering //
      expect(wrapper.find(Menu.Item).at(0).props().active).toEqual(false);
      expect(wrapper.find(Menu.Item).at(1).props().active).toEqual(true);
      expect(wrapper.find(Menu.Item).at(2).props().active).toEqual(false);
      expect(wrapper.find(Menu.Item).at(3).props().active).toEqual(false);
      // assert correct client route //
      expect(wrapper.find(Router).props().history.location.pathname).toEqual(AdminProductRoutes.CREATE_ROUTE);
    });
    it("Should correctly respond to 'Manage' 'Menu.Item' click", () => {
      window.scrollTo = jest.fn;
      const menuItems = wrapper.find(Menu.Item)
      menuItems.at(2).simulate("click");
      wrapper.update();
      // assert correct rendering //
      expect(wrapper.find(Menu.Item).at(0).props().active).toEqual(false);
      expect(wrapper.find(Menu.Item).at(1).props().active).toEqual(false);
      expect(wrapper.find(Menu.Item).at(2).props().active).toEqual(true);
      expect(wrapper.find(Menu.Item).at(3).props().active).toEqual(false);
      // assert correct client route //
      expect(wrapper.find(Router).props().history.location.pathname).toEqual(AdminProductRoutes.MANAGE_ROUTE);
    })
    it("Should correctly respond to 'View Sorted' 'Menu.Item' click", () => {
      window.scrollTo = jest.fn;
      const menuItems = wrapper.find(Menu.Item)
      menuItems.at(3).simulate("click");
      wrapper.update();
      // assert correct rendering //
      expect(wrapper.find(Menu.Item).at(0).props().active).toEqual(false);
      expect(wrapper.find(Menu.Item).at(1).props().active).toEqual(false);
      expect(wrapper.find(Menu.Item).at(2).props().active).toEqual(false);
      expect(wrapper.find(Menu.Item).at(3).props().active).toEqual(true);
      // assert correct client route //
      expect(wrapper.find(Router).props().history.location.pathname).toEqual(AdminProductRoutes.VIEW_SORTED_ROUTE);
    })
  })
})