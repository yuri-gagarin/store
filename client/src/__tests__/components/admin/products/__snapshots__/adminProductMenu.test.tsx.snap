// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminProductMenu component render tests Default component render Should properly render and match snapshot 1`] = `
<withRouter(AdminProductsMenu)
  dispatch={[Function]}
>
  <AdminProductsMenu
    dispatch={[Function]}
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/admin/home/my_products",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/admin/home/my_products",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
    location={
      Object {
        "hash": "",
        "pathname": "/admin/home/my_products",
        "search": "",
        "state": undefined,
      }
    }
    match={
      Object {
        "isExact": false,
        "params": Object {},
        "path": "/",
        "url": "/",
      }
    }
  >
    <div
      className="adminProductsMenuFixed"
    >
      <Menu
        className="adminProductsMenu"
        tabular={true}
      >
        <div
          className="ui tabular adminProductsMenu menu"
        >
          <MenuItem
            active={false}
            content="View All Products"
            name="view_all"
            onClick={[Function]}
          >
            <a
              className="item"
              onClick={[Function]}
            >
              View All Products
            </a>
          </MenuItem>
          <MenuItem
            active={false}
            content="Create New Product"
            name="create"
            onClick={[Function]}
          >
            <a
              className="item"
              onClick={[Function]}
            >
              Create New Product
            </a>
          </MenuItem>
          <MenuItem
            active={false}
            content="Manage Products"
            name="manage"
            onClick={[Function]}
          >
            <a
              className="item"
              onClick={[Function]}
            >
              Manage Products
            </a>
          </MenuItem>
          <MenuItem
            active={false}
            content="View Sorted Products"
            name="sorted"
            onClick={[Function]}
          >
            <a
              className="item"
              onClick={[Function]}
            >
              View Sorted Products
            </a>
          </MenuItem>
        </div>
      </Menu>
    </div>
  </AdminProductsMenu>
</withRouter(AdminProductsMenu)>
`;
